framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -2642,33 +2642,19 @@ protected boolean checkConstructorInvocation(
             AnnotatedDeclaredType invocation,
             AnnotatedExecutableType constructor,
             NewClassTree newClassTree) {
-        AnnotatedDeclaredType returnType = (AnnotatedDeclaredType) constructor.getReturnType();
-        // When an interface is used as the identifier in an anonymous class (e.g. new Comparable()
-        // {}) the constructor method will be Object.init() {} which has an Object return type When
-        // TypeHierarchy attempts to convert it to the supertype (e.g. Comparable) it will return
-        // null from asSuper and return false for the check.  Instead, copy the primary annotations
-        // to the declared type and then do a subtyping check.
-        if (invocation.getUnderlyingType().asElement().getKind().isInterface()
-                && TypesUtils.isObject(returnType.getUnderlyingType())) {
-            final AnnotatedDeclaredType retAsDt = invocation.deepCopy();
-            retAsDt.replaceAnnotations(returnType.getAnnotations());
-            returnType = retAsDt;
-        } else if (newClassTree.getClassBody() != null) {
-            // An anonymous class invokes the constructor of its super class, so the underlying
-            // types of invocation and returnType are not the same.  Call asSuper so they are the
-            // same and the is subtype tests below work correctly
-            invocation = AnnotatedTypes.asSuper(atypeFactory, invocation, returnType);
-        }
+        // Only check the primary annotations, the type arguments are checked elsewhere.
+        Set<AnnotationMirror> returnType = constructor.getReturnType().getAnnotations();
+        Set<AnnotationMirror> invocationAnnos = invocation.getAnnotations();
 
         // The return type of the constructor (returnType) must be comparable to the type of the
         // constructor invocation (invocation).
-        if (!(atypeFactory.getTypeHierarchy().isSubtype(invocation, returnType)
-                || atypeFactory.getTypeHierarchy().isSubtype(returnType, invocation))) {
+        if (!(atypeFactory.getQualifierHierarchy().isSubtype(invocationAnnos, returnType)
+                || atypeFactory.getQualifierHierarchy().isSubtype(returnType, invocationAnnos))) {
             checker.report(
                     Result.failure(
                             "constructor.invocation.invalid",
                             constructor.toString(),
-                            invocation,
+                            invocationAnnos,
```

Comment:
How about adapting the error message in `framework/src/main/java/org/checkerframework/common/basetype/messages.properties` from
`constructor.invocation.invalid=creation of %s not allowed with given receiver;\nfound   : %s\nrequired: %s`

to something like
`constructor.invocation.invalid=creation of %s not allowed with given annotations;\nfound annotations "%s" not comparable to "%s"`